service/matching/matching_engine.go

```diff
@@ -615,6 +615,14 @@ pollLoop:
 		resp, err := e.recordWorkflowTaskStarted(ctx, requestClone, task)
 		if err != nil {
 			switch err.(type) {
+			case *serviceerror.Internal:
+				e.nonRetryableErrorsDropTask(task, taskQueueName, err)
+			case *serviceerror.DataLoss:
+				e.nonRetryableErrorsDropTask(task, taskQueueName, err)
+			case *serialization.DeserializationError:
+				e.nonRetryableErrorsDropTask(task, taskQueueName, err)
+			case *serialization.SerializationError:
+				e.nonRetryableErrorsDropTask(task, taskQueueName, err)
```

Comment:
nit: they can be grouped under one `case`.

Reply:
hmm I guess for any new error handling logic added we should be checking wrapped error as well using errors.As()

Reply:
A couple of things, in my opinion, if we want to inculcate using `errors.as()` while checking for errors:

1. Currently, after looking at the code, it doesn't stand out to me if we are wrapping errors or not since it seems like we return the error type directly (for example - return `internal.serviceError(..)`). However, I could be wrong since i ma just getting my feet dirty with `history`. 
2. I agree with conducting `errors.As()` checks since they seem to be more stricter since they do a DFS on the error chain. What I do think we should do though, in`PollWorkflowTasks` and in `PollActivityTasks`, is replacing the type assertion with `errors.As()`. This makes things simpler + better imo.

wdyt @yycptt @dnr @ShahabT ?

Reply:
oh yeah I forgot this is across grpc...